### PR TITLE
feat(#12): item system

### LINKED ISSUE #12: Item System
- [x] Item data 
  - [x] In game Object
  - [x] Icon
- [x] Implement items
  - [x] Titanium ore
  - [x] Iron ore
  - [x] Titanium ingot 
  - [x] Iron ingot

### PR DESCRIPTION
**Pull request by PatrykPaluch**
Closes #12

### CODE REVIEW

**Comment by PatrykPaluch** [comment]
Work in progress

File: Assets/Tests/EditModeTests/SerializableDictionaryTest.cs [hunk 1]
```diff
@@ -1,3 +1,4 @@
+using System;
```
Comment (Beholder1) [inline Assets/Tests/EditModeTests/SerializableDictionaryTest.cs]: Unused imports:
- using System;
- using System.Collections;
- sing UnityEngine;
- using UnityEngine.TestTools;

File: Assets/Scripts/ItemSystem/ItemInfo.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        [SerializeField]
+        private Texture2D icon;
+        [SerializeField]
+        private GameObject prefab;
+        [SerializeField]
+        private SerializableDictionary<string, string> additionalData = new();
+
+        public string Id => id;
+        public string DisplayName => displayName;
+        public Texture2D Icon => icon;
```
Comment (Beholder1) [inline Assets/Scripts/ItemSystem/ItemInfo.cs:24]: It's unused, but I guess it exists for future purposes?

File: Assets/Scripts/ItemSystem/Editor/ItemInfoInspector.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
+using System.Linq;
+using System.Reflection;
+using UnityEditor;
+using UnityEngine;
```
Comment (Beholder1) [inline Assets/Scripts/ItemSystem/Editor/ItemInfoInspector.cs]: UnityEngine is unused.

File: Assets/Scripts/ItemSystem/Editor/ItemInfoInspector.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+namespace Cosmobot.ItemSystem.Editors
```
Comment (Beholder1) [inline Assets/Scripts/ItemSystem/Editor/ItemInfoInspector.cs]: The folder name is **Editor**, not **Editors**

File: Assets/Scripts/ItemSystem/Editor/ItemInfoInspector.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+{
+    [CustomEditor(typeof(ItemInfo))]
+    public class ItemInfoInspector : Editor
+    {
+        private ItemInfo itemInfo;
+        private SerializedProperty additionalData;
```
Comment (Beholder1) [inline Assets/Scripts/ItemSystem/Editor/ItemInfoInspector.cs]: additionalData is unused.

File: Assets/Scripts/ItemSystem/Editor/ItemInfoInspector.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+            if (undeclaredConstants.Length == 0) return;
+
+            string joined = string.Join(", ", undeclaredConstants);
+            EditorGUILayout.HelpBox(
+                $"The following keys are not declared in {nameof(ItemDataConstants)}: \n{joined}",
+                MessageType.Warning);
+        }
+
+        private void InitializeDeclaredConstants()
```
Comment (Beholder1) [inline Assets/Scripts/ItemSystem/Editor/ItemInfoInspector.cs]: Method can be static.

File: Assets/Scripts/ItemSystem/ItemInfo.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        public bool Equals(ItemInfo other) => other != null && id == other.id;
+
+        public override int GetHashCode() => id.GetHashCode();
+
+        public static bool operator ==(ItemInfo left, ItemInfo right)
+        {
+            if (ReferenceEquals(left, right)) return true;
+            if (left is null) return false;
+            if (right is null) return false;
+            return left.Equals(right);
```
Comment (Beholder1) [inline Assets/Scripts/ItemSystem/ItemInfo.cs]: Those two lines can be refactored as:
`return right is not null && left.Equals(right);`

File: Assets/Tests/PlayModeTests/ItemsTest.cs [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+            const string nullableOpKey = "nullableOpKey";
+            const string newKeyOne = "newKeyOne";
+            const string newKeyTwo = "newKeyTwo";
+
+            ItemInfo itemInfo = ItemManager.Instance.GetItem(TestItemId);
+            GameObject object1 = itemInfo.InstantiateItem(Vector3.zero, Quaternion.identity);
+            Item item = object1.GetComponent<Item>();
+
+            // operator [] not found exceptions
+            Assert.Throws<KeyNotFoundException>(() => { bool x = item.BoolValue[nonExistingKey]; });
```
Comment (Beholder1) [inline Assets/Tests/PlayModeTests/ItemsTest.cs]: It does nothing, jus asigns value to a variable.

File: Assets/Tests/PlayModeTests/ItemsTest.cs [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            item.SetIntValue(nullableOpKey, item.GetIntValue(nonExistingKey) + 2);
+            Assert.IsNull(item.GetIntValue(nullableOpKey));
+
+            // conversion
+            item.SetFloatValue(newKeyOne, 3.14f);
+            Assert.AreEqual(3.14f, item.GetFloatValue(newKeyOne));
+            Assert.AreEqual(3.14f, item.FloatValue[newKeyOne]);
+            Assert.AreEqual("3.14", item.StringValue[newKeyOne]);
+            Assert.IsNull(item.GetIntValue(newKeyOne));
+            Assert.Throws<FormatException>(() => { int x = item.IntValue[newKeyOne]; });
```
Comment (Beholder1) [inline Assets/Tests/PlayModeTests/ItemsTest.cs]: Same as above.

File: Assets/Tests/PlayModeTests/ItemsTest.cs [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+            item.BoolValue[newKeyTwo] = true;
+            Assert.AreEqual(true, item.BoolValue[newKeyTwo]);
+            Assert.AreEqual("true", item.StringValue[newKeyTwo]);
+            Assert.IsNull(item.GetIntValue(newKeyTwo));
+            Assert.IsNull(item.GetFloatValue(newKeyTwo));
+            Assert.Throws<FormatException>(() => { int x = item.IntValue[newKeyTwo]; });
```
Comment (Beholder1) [inline Assets/Tests/PlayModeTests/ItemsTest.cs]: Same as above.

File: Assets/Tests/PlayModeTests/ItemsTest.cs [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+            Assert.Throws<FormatException>(() => { float x = item.FloatValue[newKeyTwo]; });
```
Comment (Beholder1) [inline Assets/Tests/PlayModeTests/ItemsTest.cs]: Same as above.

File: Assets/Scripts/Editor/SerializableDictionaryPropertyDrawer.cs [hunk 12]
```diff
@@ -0,0 +1,121 @@
+using UnityEditor;
+using UnityEngine;
+
+namespace Cosmobot.Editors
```
Comment (Beholder1) [inline Assets/Scripts/Editor/SerializableDictionaryPropertyDrawer.cs]: Folder name is **Editor**, not **Editors**

File: Assets/Scripts/Editor/SerializableDictionaryPropertyDrawer.cs [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+{
+    [CustomPropertyDrawer(typeof(SerializableDictionary<,>))]
+    public class SerializableDictionaryPropertyDrawer : PropertyDrawer
+    {
+        private const string PROP_KEYS = "keys";
+        private const string PROP_VALUES = "values";
+
+        private const float hPadding = 4;
+        private const float vPadding = 2;
```
Comment (Beholder1) [inline Assets/Scripts/Editor/SerializableDictionaryPropertyDrawer.cs]: Values should be in Pascal case:
`PropKeys
PropValues
HPadding
VPadding`

File: Assets/Scripts/ItemSystem/ItemInfo.cs (hunk 2 above)
Comment (PatrykPaluch) [reply Assets/Scripts/ItemSystem/ItemInfo.cs:24]: yep

File: Assets/Scripts/ItemSystem/ItemInfo.cs (hunk 7 above)
Comment (PatrykPaluch) [reply Assets/Scripts/ItemSystem/ItemInfo.cs]: i think `if if return` is more readable, but ok

**Review by Beholder1 — APPROVED** [review]

**Pull request merged**